designs/0032-stan-output-formats.md

```diff
@@ -0,0 +1,295 @@
+allowing for lists, maps, and nested lists and maps.
+
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+We propose to break down the information from the Stan inference algorithms into a series of output streams
+where each stream handles one kind of information and whose format is appropriate for further processing.
+Tabular data will be output either in CSV format (human readable) or using the Apache Arrow binary formats.
+For hierarchical data we will use JSON format.
+
+Tabular information includes:
+
+- The output from the Stan model class `write_array` function: the computed values for all Stan model variables on the constrained scale.
+- The Stan model parameter variables on the unconstrained scale returned by the `log_prob` function.
+- The inference algorithm state at each step of the iterative algorithm.
+
+Structured information includes:
+
+- The HMC tuning parameters:  metric and stepsize.  The HMC metric is a symmetric positive definite matrix, either a diagonal or a dense matrix.
+The former is reported as just the vector that is matrix diagonal, the latter is reported as the full matrix.  The HMC stepsize is a scalar.
+
+- Stan model variables schema: a list of all model variables including: block, name, type, shape, size, output column name(s)
+
+- Information about the inference run:  algorithm configuration, Stan version, model compiler version and compile options, and timing information.
+
+## Implementation
+
+To improve and generalize the outputs and output formats, we propose to:
+
+- implement a new set of algorithm-specific output handlers which
+will be passed in to the `stan::services` layer wrapper function and which
+will manage the necessary callback writer objects.
+
+- decouple the output format from the writer through the use of formatter callbacks.
+For tabular data we will implement CSV and Apache Arrow formatters.
+For extra-tabular structured data we will implement a JSON formatter.
+
+The CSV structured data formatter corresponds to the current set of methods on the base callbacks writer class.
+The JSON and Apache Arrow formatters would need to be able to handle lists and dictionaries, for the former,
+and scalar and vectors of ints and doubles for the latter.
+
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+The more different kinds of output files, more effort required on the part of the user
+to configure and manage these outputs.
+
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+- Why is this design the best in the space of possible designs?
+
+The proposed refactoring uses the writer callbacks and extends them.
+
+- What other designs have been considered and what is the rationale for not choosing them?
+
+For the output formats, we considered using Google's protocol buffers.CmdStan output have
+This is a Google library which has been open-source since 2008.
```

Comment:
This sentence seems like it got mangled somehow

Reply:
thanks.